common/crypto/platform/openssl/crypto.cpp

```diff
@@ -147,11 +182,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
+
+	ENGINE_load_builtin_engines();
+	auto engine = EnginePtr(ENGINE_by_id(args.ssl_engine.c_str()), engine_free_func);
+
+	if (engine == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to get the " + args.ssl_engine
+				+ " engine. No engine with the ID found: " + GetOpenSSLErrorMessage()));
+	}
+	log::Debug("Loaded the HSM engine successfully!");
+
+	int res = ENGINE_init(engine.get());
+	if (not res) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to initialise the hardware security module (HSM): "
+				+ GetOpenSSLErrorMessage()));
+	}
+	log::Debug("Successfully initialised the HSM engine");
+
+	auto private_key = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>(
+		ENGINE_load_private_key(
+			engine.get(),
+			args.private_key_path.c_str(),
+			(UI_METHOD *) NULL,
+			NULL /*callback_data */),
+		pkey_free_func);
+	if (private_key == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to load the private key from the hardware security module: "
+				+ GetOpenSSLErrorMessage()));
+	}
+	log::Debug("Successfully loaded the private key from the HSM Engine: " + args.ssl_engine);
+
+	return PrivateKey(std::move(private_key), std::move(engine));
+}
+#endif // MENDER_CRYPTO_OPENSSL_LEGACY
+
+#ifndef MENDER_CRYPTO_OPENSSL_LEGACY
+ExpectedPrivateKey PrivateKey::LoadFromHSM(const Args &args) {
+	log::Debug("Loading the private key from HSM");
+
+	auto default_provider = ProviderPtr(OSSL_PROVIDER_load(NULL, "default"), OSSL_PROVIDER_unload);
+	if (default_provider == nullptr) {
+		return expected::unexpected(
+			MakeError(SetupError, "default provider load error: " + GetOpenSSLErrorMessage()));
+	}
+
+	auto hsm_provider =
+		ProviderPtr(OSSL_PROVIDER_load(NULL, args.ssl_engine.c_str()), OSSL_PROVIDER_unload);
+	if (hsm_provider == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError, args.ssl_engine + " provider load error: " + GetOpenSSLErrorMessage()));
+	}
+
+	int ret {OSSL_PROVIDER_available(NULL, args.ssl_engine.c_str())};
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError, args.ssl_engine + " provider not available: " + GetOpenSSLErrorMessage()));
+	}
+
+	auto ctx = unique_ptr<OSSL_STORE_CTX, int (*)(OSSL_STORE_CTX *)>(
+		OSSL_STORE_open(args.private_key_path.c_str(), NULL, NULL, NULL, NULL), OSSL_STORE_close);
+
+	if (ctx == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"OSSL_STORE_OPEN: Failed to load the private key from the hardware security module: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	// Go through all objects in the context till we find the first private key
+	while (not OSSL_STORE_eof(ctx.get())) {
+		OSSL_STORE_INFO *info = OSSL_STORE_load(ctx.get());
+
+		if (info == NULL) {
+			log::Error(
+				"Failed to load the store info from the hardware security module: trying the next object in the context: "
+				+ GetOpenSSLErrorMessage());
+		}
```

Comment:
Missing `continue`? I see there is one below, but that's after `info` has already been used, so is it correct?

Reply:
The other continue is for the case where we don't get a `private-key` from the `info` struct, but some other type.

Reply:
Ok, so we do need it here as well?

Reply:
Yes :) thank you!